Decrement workspace_entity_counters on cascade deletes

tyler-ball opened the pull request:
### Summary

Pulling the non-ee changes from https://github.com/Kong/kong-ee/pull/2392 to here. See that PR for an explanation of the issue and my proposed solution.

### Full changelog

☝️ 

### Issues resolved

Fix FTI-1847

tyler-ball commented on spec/01-unit/01-db/01-schema/01-schema_spec.lua:
```diff
@@ -3787,6 +3787,75 @@ describe("schema", function()
+      assert.is.Truthy(Entities3)
+      local constraints = Entities1:get_constraints()
+
+      assert.are.same({
```
This test is failing because the `constraints` list has the items in a different order - is there some way to assert unordered arrays are the same? What is the best pattern to do that?

javierguerragiraldez replied:
i typically sort the array before comparing.

javierguerragiraldez commented on kong/db/dao/init.lua:
```diff
@@ -1297,6 +1299,15 @@ function DAO:delete(primary_key, options)
+    if not cascade_entity then
+      return nil, tostring(err_t), err_t
+    end
+  end
```
this is the same as lines 912-920, right?  can it be factored into a function?

tyler-ball replied:
Done

dndx commented:
Ping @tyler-ball ^^

tyler-ball commented:
@javierguerragiraldez @dndx I've addressed Javier's comments, which were pretty minor. Any feedback about the overall approach to solving the original issue (not decrementing counters)?

tyler-ball commented:
```
        ERROR  spec/01-unit/01-db/04-dao_spec.lua:461: DAO delete deletes the entity and cascades the delete notifications
./kong/db/dao/init.lua:1433: attempt to index global 'kong' (a nil value)

stack traceback:
	./kong/db/dao/init.lua:1433: in function 'row_to_entity'
	./kong/db/dao/init.lua:1008: in function 'select'
	./kong/db/dao/init.lua:1269: in function 'delete'
	spec/01-unit/01-db/04-dao_spec.lua:493: in function <spec/01-unit/01-db/04-dao_spec.lua:461>
```
I don't get this error in `kong-ee`. I tried `local kong = kong` which I see in some other files but that did not help. Whats the best way to get this global reference imported correctly?

gszr commented on kong/db/dao/init.lua:
```diff
@@ -909,14 +918,14 @@ local function generate_foreign_key_methods(schema)
         if not entity then
           return nil, tostring(err_t), err_t
         end
+        err_t = run_hook_cascade_delete("dao:delete_by:post", cascade_entries, options)
```
A bit of historical context.
The way this 'hook system' was designed to work:
1. Core has 'named hooking points' - marked by `run_hook`
2. Enterprise leverages it by registering named hooks (which run in places specified by 1.)

Is there a need for a new hooking point here, where `run_hook_cascade_delete` is called? If so, can it be made like other ones, without extra logic, and the hook registered elsewhere?

tyler-ball replied:
I could introduce a `dao:delete_by:post_cascade` hook. I find it difficult to answer whether we "need" to.

We could also modify the existing `dao:delete_by:post` hook to accept the cascade_entries - this was one of my original implementations. I do not understand the benefit of adding a new hook as opposed to modifying the existing one. If the hooks are there to support EE, and I know the required EE logic is that we decrement the `workspace_entity_counters` for cascade_entries, why not modify the existing hook? Why introduce additional complexity with a completely new hook that needs to be called right after the existing hook?

I understand what we need to do - perform the same decrementing logic for the parent entity as well as cascade delete entities. It seems to me we keep circling around different implementations that, IMO, are coding style differences. I defer to developers with more tenure on style issues which is why I don't feel like I can answer whether we need to introduce a new hook.

gszr replied:
@tyler-ball and I will move this discussion over to Slack / Zoom

gszr commented:
> ```
>         ERROR  spec/01-unit/01-db/04-dao_spec.lua:461: DAO delete deletes the entity and cascades the delete notifications
> ./kong/db/dao/init.lua:1433: attempt to index global 'kong' (a nil value)
> 
> stack traceback:
> 	./kong/db/dao/init.lua:1433: in function 'row_to_entity'
> 	./kong/db/dao/init.lua:1008: in function 'select'
> 	./kong/db/dao/init.lua:1269: in function 'delete'
> 	spec/01-unit/01-db/04-dao_spec.lua:493: in function <spec/01-unit/01-db/04-dao_spec.lua:461>
> ```
> 
> I don't get this error in `kong-ee`. I tried `local kong = kong` which I see in some other files but that did not help. Whats the best way to get this global reference imported correctly?

That error means the kong global wasn't initialized (as that is a unit test, with minimal environment set up). Doing `local kong = kong` won't help if `kong` is nil. 

Take a peak at `spec/01-unit/12-plugins_order_spec.lua`, for example. It has the following chunk to initialize the `kong` global:
```
    local kong_global = require "kong.global"
    _G.kong = kong_global.new()
   ...
```

gszr commented on kong/db/dao/init.lua:
```diff
@@ -630,27 +630,23 @@ local function check_upsert(self, key, entity, options, name)
+    if constraint.on_delete == "cascade" then
+      local dao = self.db.daos[constraint.schema.name]
+      local method = "each_for_" .. constraint.field_name
+      for row, err in dao[method](dao, pk, nil, show_ws_id) do
```
show_ws_id?

tyler-ball commented:
@gszr I modified the DAO to pass the `cascade_entries` as we discussed in zoom - now the hook in EE will be responsible for looping over them. I also cleaned up the commits and the commit messages.

kikito commented:
Not including this on the changelog since it is EE-related
